accepted/2024/observable_collection.md

```diff
@@ -0,0 +1,173 @@
+  isn't a behavioral change as existing code won't start raising bulk events
+  unless the user is calling the new methods on `ObservableCollection<T>`. The
+  guidance would be "use bulk operations if the data bound UI supports it,
+  otherwise don't". We would treat the lack of WPF support as a feature request
+  and let it run its course.
+  - Pro: Simplest API, easiest to implement, unblocks the ecosystem.
+  - Con: Violates our goal to not ship functionality we know won't work with
+    core in-box functionality (specifically WPF).
+
+* **Global switch**. We could add an `AppContext` switch which disables bulk
+  updates process-wide.
+  - Pro: Easy to implement, reduced complexity for the API.
+  - Con: limiting if some UI components support bulk updates and some don't.
+
+* **Per instance switch**. We could have a property on `ObservableCollection<T>`
+  that turns off bulk notifications.
+  - Pro: Allows some part of the view model to use bulk notifications when the
+    UI supports it while turning it off for the parts that don't.
+  - Con: Still limiting when the same collection is bound to multiple UI
+    components and some support bulk notifications and some don't. Also
+    complicates the core API.
+
+* **Event adaptor**. We could provide an implementation of
+  `INotifyCollectionChanged` that wraps an existing `INotifyCollectionChanged`
+  and translates bulk operations to individual events. Authors of view models
+  could expose two properties, one with the actual collection and another one
+  that returns the wrapper. In data binding, they use the property with bulk
+  notifications if the UI supports it and the other if it doesn't.
+  - Pro: Core API is simple while providing an affordance to the user to solve
+    the problem without requiring changes from the UI framework.
+  - Con: More complex state, might not play well with WPF's `CollectionView`.
+
+* **Event enumerator**. We could expose a method on
+  `NotifyCollectionChangedEventArgs` that returns an
+  `IEnumerable<NotifyCollectionChangedEventArgs>`. For single item events it
+  returns itself, otherwise it creates a new event arg per item. This requires
+  UI controls to make code changes if they don't support bulk notifications, but
+  the change is fairly targeted.
+  - Pro: Simple API
+  - Con: Still breaks consumers, just provides an easier path to adapt to the
+    new behavior without having to full support bulk notifications.
+
+* **Handling it at the CollectionView level**. WPF seems to have an indirection
+  between the data-bound UI controls and the observable collections via the
+  collection view. Maybe we could expose a property on `CollectionView` that
+  controls whether the collection view will translate bulk events to individual
+  events.
+  - Pro: Simpler API for the base, with an opt-out switch at the level of WPF.
+  - Con: Needs more investigation to determine feasibility.
```

Comment:
Con: This also only is a solution for one out of numerous platforms.

Reply:
```suggestion
  - Con: Needs more investigation to determine feasibility, would only be helpful for WPF
```